packages/console/src/pages/UserDetails/index.tsx

```diff
@@ -1,32 +1,101 @@
 import { User } from '@logto/schemas';
-import React from 'react';
+import React, { useEffect, useState } from 'react';
+import { useController, useForm } from 'react-hook-form';
+import { toast } from 'react-hot-toast';
 import { useTranslation } from 'react-i18next';
 import { useParams } from 'react-router-dom';
 import useSWR from 'swr';
 
 import BackLink from '@/components/BackLink';
+import Button from '@/components/Button';
 import Card from '@/components/Card';
+import CodeEditor from '@/components/CodeEditor';
 import CopyToClipboard from '@/components/CopyToClipboard';
+import FormField from '@/components/FormField';
 import ImagePlaceholder from '@/components/ImagePlaceholder';
-import { RequestError } from '@/hooks/use-api';
+import TabNav, { TabNavLink } from '@/components/TabNav';
+import TextInput from '@/components/TextInput';
+import useApi, { RequestError } from '@/hooks/use-api';
+import { safeParseJson } from '@/utilities/json';
 
 import CreateSuccess from './components/CreateSuccess';
 import * as styles from './index.module.scss';
 
+type FormData = {
+  primaryEmail: string;
+  primaryPhone: string;
+  username: string;
+  name: string;
+  avatar: string;
+  roles: string;
+  customData: string;
+};
+
 const UserDetails = () => {
   const { id } = useParams();
   const { t } = useTranslation(undefined, { keyPrefix: 'admin_console' });
 
-  const { data, error } = useSWR<User, RequestError>(id && `/api/users/${id}`);
+  const { data, error, mutate } = useSWR<User, RequestError>(id && `/api/users/${id}`);
   const isLoading = !data && !error;
 
+  const { handleSubmit, register, control, reset } = useForm<FormData>();
+  const [submitting, setSubmitting] = useState(false);
+  const {
+    field: { onChange, value },
+  } = useController({ name: 'customData', control, rules: { required: true } });
+  const api = useApi();
+
+  useEffect(() => {
+    if (!data) {
+      return;
+    }
+    reset({
+      primaryEmail: data.primaryEmail ?? '',
+      primaryPhone: data.primaryPhone ?? '',
+      username: data.username ?? '',
+      name: data.name ?? '',
+      avatar: data.avatar ?? '',
+      roles: data.roleNames.join(','),
+      customData: JSON.stringify(data.customData, null, 2),
+    });
+  }, [data, reset]);
+
+  const onSubmit = handleSubmit(async (formData) => {
+    if (!data || submitting) {
+      return;
+    }
+
+    const customData = safeParseJson(formData.customData);
+
+    if (!customData) {
+      toast.error(t('user_details.custom_data_invalid'));
+
+      return;
+    }
+
+    const payload: Partial<User> = {
+      name: formData.name || undefined,
+      avatar: formData.avatar || undefined,
```

Comment:
why `|| undefined`? Will `name` and `avatar` be `null` or something else negative?

Reply:
updated